src/Function.cpp

```diff
@@ -248,6 +299,129 @@ Function::Function(const std::string &n) : contents(new FunctionContents) {
     contents.ptr->name = n;
 }
 
+// Deep-copy UpdateDefinition from 'src' to 'dst'
+void deep_copy_update_definition_helper(UpdateDefinition &dst,
+                                        const UpdateDefinition &src,
+                                        DeepCopyMap &copied_map) {
+    dst.values = src.values;
+    dst.args = src.args;
+    dst.schedule = src.schedule.deep_copy(copied_map);
+
+    // UpdateDefinition's domain is the same as the one pointed by its schedule
+    internal_assert(src.schedule.reduction_domain().same_as(src.domain))
+        << "UpdateDefinition should point to the same reduction domain as its schedule\n";
+    // We don't need to deep-copy the reduction domain since we've already done
+    // it when deep-copying the schedule above
+    dst.domain = dst.schedule.reduction_domain();
+}
+
+// Deep-copy ExternFuncArgument from 'src' to 'dst'
+void deep_copy_extern_func_argument_helper(ExternFuncArgument &dst,
```

Comment:
Mixing in, out, and in-out parameters in this way is hard to read and kinda unexpected. Could we just make DeepCopyMap a real class (to maintain the state) and make these methods on it:

    void DeepCopyMap::copy_extern_func_argument(const ExternFuncArgument &src, ExternFuncArgument &dst)

or maybe better, return as a result:

    ExternFuncArgument DeepCopyMap::copy_extern_func_argument(const ExternFuncArgument &src)